doc/contributing/coding-conventions.chapter.md

```diff
@@ -512,3 +512,73 @@ If you do need to do create this sort of patch file, one way to do so is with gi
     ```ShellSession
     $ git diff > nixpkgs/pkgs/the/package/0001-changes.patch
     ```
+
+## Package tests {#sec-package-tests}
+
+Tests are important to ensure quality and make reviews and automatic updates easy.
+
+Nix package tests are a lightweight alternative to [NixOS module tests](https://nixos.org/manual/nixos/stable/#sec-nixos-tests). They can be used to create simple integration tests for packages while the module tests are used to test services or programs with a graphical user interface on a NixOS VM. Unittests that are included in the source code of a package should be executed in the `checkPhase`.
```

Comment:
> module tests are used to test services or programs with a graphical user interface on a NixOS VM

We can handle graphical test just fine using Xvfb. The NixOS tests are primarily integration tests that depend on other services or to test NixOS modules themselves. `passthru.tests` should really be preferred whenever possible since the NixOS tests really are heavy.

Reply:
great. do you have an example?
i need it for https://github.com/NixOS/nixpkgs/pull/120453#discussion_r619632124
can it replace https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/shattered-pixel-dungeon.nix ?

Reply:
The invocation is

https://github.com/NixOS/nixpkgs/blob/b7e08de1e9e9a3e484fd8952151c59571e5c479c/pkgs/development/libraries/libdazzle/default.nix#L26-L28

Sometimes the dbus session part can be omitted.

>  can it replace https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/shattered-pixel-dungeon.nix ?

Yeah, that should not be NixOS test.

But xvfb-run just creates a virtual frame buffer so that you can start graphical programs – you will still need to write your own tests or use upstream ones.  We should probably create some helpers for passthru tests so that we can use tesseract [like in the NixOS test](https://github.com/NixOS/nixpkgs/blob/b7e08de1e9e9a3e484fd8952151c59571e5c479c/nixos/lib/test-driver/test-driver.py#L208-L236).

Reply:
Shouldn't it be `installCheckPhase` is you want to e.g. test the final result? (IIRC that was wan of the ideas behind the NixCon 2017 talk as well).

Reply:
Clarification on that is added with https://github.com/NixOS/nixpkgs/pull/119731.

Reply:
@jtojnar thanks!

>We should probably create some helpers for passthru tests so that we can use tesseract like in the NixOS test.

that's probably what https://github.com/NixOS/nixpkgs/pull/36842 tried, but it was rejected because it duplicated nixos tests... maybe it can be generalized somehow so it works for both